### DISCUSSION TITLE
What are glob patterns and how to use them?

### DISCUSSION THREAD

**Discussion by NebelNidas** [Kira-NT/mc-publish#30] [Q&A] [answered]
Can we get a more comprehensive explanation of the file selector syntax? Or a link to one? Because I'm trying to do something like this: 

```json
"files-primary": "toomanybinds-wrapper/build/libs/((*+1.19) && !(*-@(dev|sources))).jar"
```
, so an inclusive condition plus an exclusive one combined, and I can't find out if it works without uploading (potentially wrong) files to CurseForge/Modrinth 😅 Also, is it possible to use regex?

**Comment by Kira-NT** [comment, marked as answer]
Hi! :)

> Can we get a more comprehensive explanation of the file selector syntax?

Please, take a look at the `Inputs` section of README a little bit more carefully:

| Name | Description | Default value | Examples |
|------|-------------|---------------|----------|
| [files](https://github.com/Kir-Antipov/mc-publish#user-content-files) | A **<ins>glob</ins>** of the file(s) to upload | ❌ | `build/libs/*.jar` |
| [files-primary](https://github.com/Kir-Antipov/mc-publish#user-content-files-primary) | A **<ins>glob</ins>** of the primary files to upload | `build/libs/!(*-@(dev\|sources)).jar` | `build/libs/!(*-@(dev\|sources)).jar` |
| [files-secondary](https://github.com/Kir-Antipov/mc-publish#user-content-files-secondary) | A **<ins>glob</ins>** of the secondary files to upload | `build/libs/*-@(dev\|sources).jar` | `build/libs/*-@(dev\|sources).jar` |

It clearly states that `files`, `files-primary`, and `files-secondary` inputs use [`glob`](https://en.wikipedia.org/wiki/Glob_(programming)) patterns. Of course, I did not invent ~~a wheel~~ yet another pattern language :)

------ 

> so an inclusive condition plus an exclusive one combined

If I understand your intentions correctly, you need something like this:

```yaml
files-primary: toomanybinds-wrapper/build/libs/*+1.19!(*-@(dev|sources)).jar
```

Just like with regular expressions, you do not need to specify a conjunction explicitly.

------

> I can't find out if it works without uploading (potentially wrong) files to CurseForge/Modrinth 😅

Don't worry, there are a lot of online tools that can help you test your glob patterns. Personally, I like the one made by [DigitalOcean, where you can test that glob I wrote, and learn more about globs in general](https://www.digitalocean.com/community/tools/glob?comments=true&glob=toomanybinds-wrapper%2Fbuild%2Flibs%2F%2A%2B1.19%21%28%2A-%40%28dev%7Csources%29%29.jar&matches=false&tests=toomanybinds-wrapper%2Fbuild%2Flibs%2Ftoomanybinds-wrapper%2B1.19.jar&tests=toomanybinds-wrapper%2Fbuild%2Flibs%2Ftoomanybinds-wrapper%2B1.19-too.jar&tests=toomanybinds-wrapper%2Fbuild%2Flibs%2Ftoomanybinds-wrapper%2B1.19-dev.jar&tests=toomanybinds-wrapper%2Fbuild%2Flibs%2Ftoomanybinds-wrapper%2B1.19-sources.jar&tests=toomanybinds-wrapper%2Fbuild%2Flibs%2Ftoomanybinds-wrapper%2B1.18.jar&tests=toomanybinds-wrapper%2Fbuild%2Flibs%2Ftoomanybinds-wrapper%2B1.18-too.jar&tests=toomanybinds-wrapper%2Fbuild%2Flibs%2Ftoomanybinds-wrapper%2B1.18-dev.jar&tests=toomanybinds-wrapper%2Fbuild%2Flibs%2Ftoomanybinds-wrapper%2B1.18-sources.jar).

-------

> Also, is it possible to use regex?

No, it's not and it will likely never be possible. Support of more than one pattern language may be confusing for users, and it is also unnecessary. Globs are far more suitable for file selection than more generic regular expressions. This syntax is the de facto standard for such tasks wherever you look, from `.gitignore` to GitHub Actions like `action/upload-artifact` :)
